sphinx/ext/autodoc/__init__.py

```diff
@@ -1341,6 +1356,7 @@ def get_user_defined_function_or_method(obj: Any, attr: str) -> Any:
         # the signature from, so just pass the object itself to our hook.
         self.env.app.emit('autodoc-before-process-signature', self.object, False)
         try:
+            self._signature_method = None
             return inspect.signature(self.object, bound_method=False)
```

Comment:
It's a bit ugly how we can't just return it here. Perhaps we should allow `format_args` to return multiple lines as well, and then we would have access to the return value?

Or as a middle ground, change this to
```suggestion
            return None, inspect.signature(self.object, bound_method=False)
```
and change the line below as
```diff
-sig = self._get_signature()
+sig_method, sig = self._get_signature()
+self._signature_method = sig_method  # TODO: get this to `format_signature` in a tidier way
```

Reply:
Sounds good to change retval of `_get_signature()`. I'm okay if we don't change interface of public method (`format_signature()` and `format_args()`). But it causes `_get_signature()` is called twice in both `format_signature()` and `format_args()`.
```
def format_signature(self, **kwargs):
    sig_method, sig = self._get_signature()
    if sig_method and self.analyzer:
        ...
    else:
        overloaded = False
        sig = super().format_signature()

   ...
```

Reply:
Note: it would be much better to return a tuple of the target class (self.object or metaclass), method name and signature string.